resctrl/utils.go

```diff
@@ -169,108 +194,102 @@ func getAllProcessThreads(threadFiles []os.FileInfo) ([]int, error) {
 	return processThreads, nil
 }
 
-func findControlGroup(pids []string) (string, error) {
+// findGroup returns the path of a control/monitoring group in which the pids are.
+func findGroup(group string, pids []string, includeGroup bool, exclusive bool) (string, error) {
 	if len(pids) == 0 {
 		return "", fmt.Errorf(noPidsPassedError)
 	}
-	availablePaths, err := filepath.Glob(filepath.Join(rootResctrl, "*"))
+
+	availablePaths := make([]string, 0)
+	if includeGroup {
+		availablePaths = append(availablePaths, group)
+	}
+
+	filenames, err := ioutil.ReadDir(group)
+	for _, file := range filenames {
+		if _, ok := groupDirectories[file.Name()]; !ok {
+			availablePaths = append(availablePaths, filepath.Join(group, file.Name()))
+		}
+	}
 	if err != nil {
-		return "", err
+		return "", fmt.Errorf("couldn't obtain groups paths: %w", err)
 	}
 
 	for _, path := range availablePaths {
-		switch path {
-		case filepath.Join(rootResctrl, cpusFileName):
-			continue
-		case filepath.Join(rootResctrl, cpusListFileName):
-			continue
-		case filepath.Join(rootResctrl, modeFileName):
-			continue
-		case filepath.Join(rootResctrl, sizeFileName):
-			continue
-		case filepath.Join(rootResctrl, infoDirName):
-			continue
-		case filepath.Join(rootResctrl, monDataDirName):
-			continue
-		case filepath.Join(rootResctrl, monGroupsDirName):
-			continue
-		case filepath.Join(rootResctrl, schemataFileName):
-			continue
-		case filepath.Join(rootResctrl, tasksFileName):
-			inGroup, err := arePIDsInControlGroup(rootResctrl, pids)
-			if err != nil {
-				return "", err
-			}
-			if inGroup {
-				return rootResctrl, nil
-			}
-		default:
-			inGroup, err := arePIDsInControlGroup(path, pids)
-			if err != nil {
-				return "", err
-			}
-			if inGroup {
-				return path, nil
-			}
+		groupFound, err := arePIDsInGroup(path, pids, exclusive)
+		if err != nil {
+			return "", err
+		}
+		if groupFound {
+			return path, nil
 		}
 	}
 
-	return "", fmt.Errorf("there is no available control group")
+	return "", nil
 }
 
-func arePIDsInControlGroup(path string, pids []string) (bool, error) {
+// arePIDsInGroup returns true if all of the pids are within control group.
+func arePIDsInGroup(path string, pids []string, exclusive bool) (bool, error) {
 	if len(pids) == 0 {
-		// No container pids passed.
 		return false, fmt.Errorf("couldn't obtain pids from %q path: %v", path, noPidsPassedError)
 	}
 
-	tasksFile, err := ioutil.ReadFile(filepath.Join(path, "tasks"))
+	tasks, err := readTasksFile(filepath.Join(path, tasksFileName))
 	if err != nil {
-		return false, fmt.Errorf("couldn't obtain pids from %q path: %w", path, err)
-	}
-
-	if len(tasksFile) == 0 {
-		return false, nil
+		return false, err
 	}
 
-	tasks := readTasksFile(tasksFile)
-
-	for _, pid := range pids {
+	any := false
+	for _, pidString := range pids {
+		pid, err := strconv.Atoi(pidString)
```

Comment:
conversion for every found pid (in both container and potenial mon_group) costs more than just keeping strings (and comparing strings)